File: Spring Boot(Backend)/mkbankspring/src/main/java/com/emranhss/mkbankspring/dto/TransactionDTO.java
package com.emranhss.mkbankspring.dto;

import java.util.Date;

public class TransactionDTO {
    private Long id;

    // Embed full account details
    private AccountsDTO account;          // sender account
    private AccountsDTO receiverAccount;  // receiver account

    private String type; // DEBIT or CREDIT
    private String transactionType;
    private double amount;
    private Date transactionTime;
    private String description;

    // Bill payment info
    private String companyName;
    private String accountHolderBillingId;

    public TransactionDTO() {}

    public TransactionDTO(Long id, AccountsDTO account, AccountsDTO receiverAccount, String type,
                          String transactionType, double amount, Date transactionTime, String description,
                          String companyName, String accountHolderBillingId) {
        this.id = id;
        this.account = account;
        this.receiverAccount = receiverAccount;
        this.type = type;
        this.transactionType = transactionType;
        this.amount = amount;
        this.transactionTime = transactionTime;
        this.description = description;
        this.companyName = companyName;
        this.accountHolderBillingId = accountHolderBillingId;
    }

    // Getters & Setters
    public Long getId() {
        return id;
    }

    public void setId(Long id) {
        this.id = id;
    }

    public AccountsDTO getAccount() {
        return account;
    }

    public void setAccount(AccountsDTO account) {
        this.account = account;
    }

    public AccountsDTO getReceiverAccount() {
        return receiverAccount;
    }

    public void setReceiverAccount(AccountsDTO receiverAccount) {
        this.receiverAccount = receiverAccount;
    }

    public String getType() {
        return type;
    }

    public void setType(String type) {
        this.type = type;
    }

    public String getTransactionType() {
        return transactionType;
    }

    public void setTransactionType(String transactionType) {
        this.transactionType = transactionType;
    }

    public double getAmount() {
        return amount;
    }

    public void setAmount(double amount) {
        this.amount = amount;
    }

    public Date getTransactionTime() {
        return transactionTime;
    }

    public void setTransactionTime(Date transactionTime) {
        this.transactionTime = transactionTime;
    }

    public String getDescription() {
        return description;
    }

    public void setDescription(String description) {
        this.description = description;
    }

    public String getCompanyName() {
        return companyName;
    }

    public void setCompanyName(String companyName) {
        this.companyName = companyName;
    }

    public String getAccountHolderBillingId() {
        return accountHolderBillingId;
    }

    public void setAccountHolderBillingId(String accountHolderBillingId) {
        this.accountHolderBillingId = accountHolderBillingId;
    }
}

File: Flutter(Frontend)/mk_bank_project/lib/dto/transactiondto.dart


import 'package:mk_bank_project/dto/accountsdto.dart';

class TransactionDTO {
  final int? id;
  final AccountsDTO account; // পূর্ণ Account object
  final AccountsDTO? receiverAccount; // পূর্ণ Receiver Account object (nullable)
  final String? type; // DEBIT or CREDIT
  final String? transactionType; // e.g., TRANSFER, BILL_PAYMENT
  final double? amount;
  final String? transactionTime;
  final String? description;
  final String? companyName;
  final String? accountHolderBillingId;

  // Statement Page এর জন্য অতিরিক্ত ফিল্ড
  final double? runningBalance;

  TransactionDTO({
    this.id,
    required this.account,
    this.receiverAccount,
    this.type,
    this.transactionType,
    this.amount,
    this.transactionTime,
    this.description,
    this.companyName,
    this.accountHolderBillingId,
    this.runningBalance, // runningBalance এখানে রাখুন
  });

  // API রেসপন্স থেকে অবজেক্ট তৈরি করা
  factory TransactionDTO.fromJson(Map<String, dynamic> json) {
    return TransactionDTO(
      id: json['id']?.toInt(),
      // গুরুত্বপূর্ণ: nested JSON কে AccountsDTO তে কনভার্ট করা
      account: AccountsDTO.fromJson(json['account']),
      receiverAccount: json['receiverAccount'] != null
          ? AccountsDTO.fromJson(json['receiverAccount'])
          : null,
      type: json['type'],
      transactionType: json['transactionType'],
      amount: (json['amount'] is int) ? (json['amount'] as int).toDouble() : json['amount']?.toDouble(),
      transactionTime: json['transactionTime'],
      description: json['description'],
      companyName: json['companyName'],
      accountHolderBillingId: json['accountHolderBillingId'],
      // runningBalance API থেকে আসে না, তাই এখানে null থাকবে
      runningBalance: null,
    );
  }

  // runningBalance আপডেট করার জন্য copyWith মেথড (StatementPage এ ব্যবহৃত)
  TransactionDTO copyWith({
    double? runningBalance,
  }) {
    return TransactionDTO(
      id: id,
      account: account,
      receiverAccount: receiverAccount,
      type: type,
      transactionType: transactionType,
      amount: amount,
      transactionTime: transactionTime,
      description: description,
      companyName: companyName,
      accountHolderBillingId: accountHolderBillingId,
      runningBalance: runningBalance ?? this.runningBalance,
    );
  }
}